internal/uvm/scsi.go

```diff
@@ -50,6 +54,72 @@ func (sm *SCSIMount) Release(ctx context.Context) error {
 	return nil
 }
 
+// If a SCSI mount is read only then we should simply add it to the uvm config. But if it
+// is a scratch layer (i.e writeable mount) then we should make a copy of it.
+func (sm *SCSIMount) Clone(ctx context.Context, vm *UtilityVM, cd *CloneData) (interface{}, error) {
+	var (
+		dstVhdPath string = sm.HostPath
+		err        error
+		dir        string
+	)
+	conStr := fmt.Sprintf("%d", sm.Controller)
+	lunStr := fmt.Sprintf("%d", sm.LUN)
+	if !sm.IsReadOnly {
+		// Copy this scsi disk
+		// TODO(ambarve): This can be a SCSI mount that belongs to some container
+		// which is being automatically cloned here as a part of UVM cloning
+		// process. We will receive a request for creation of this container
+		// later on which will specify the storage path for this container.
+		// However, that storage location is not available now so we just use
+		// the storage of the uvm instead. Find a better way for handling this.
+
+		// For the scratch VHD of the VM clone it in the scratch folder
+		if sm.Controller == 0 && sm.LUN == 0 {
+			dir = cd.scratchFolder
+		} else {
+			dir, err = ioutil.TempDir(cd.scratchFolder, fmt.Sprintf("clone-mount-%d-%d", sm.Controller, sm.LUN))
+			if err != nil {
+				return nil, fmt.Errorf("Error while creating directory for scsi mounts of clone vm: %s", err)
+			}
+		}
+
+		// copy the VHDX
+		dstVhdPath = filepath.Join(dir, filepath.Base(sm.HostPath))
+		log.G(ctx).Debug(fmt.Sprintf("cloning scsi mount with hostPath: %s, Controller: %d, LUN: %d to: %s", sm.HostPath, sm.Controller, sm.LUN, dstVhdPath))
+
+		if err = copyfile.CopyFile(ctx, sm.HostPath, dstVhdPath, true); err != nil {
```

Comment:
Do we have any info on how much faster cloning is over normal pod/container creation when we have a lot of vhds to copy?

Reply:
We haven't done that kind of performance testing yet. Ideally you should not do a lot of processing in a template container and so these scratch VHDs should be very small.